Dockerfile

```diff
@@ -1,38 +1,76 @@
+ARG RUST_VER=1.65
+ARG VECTORSCAN_VER=5.4.8
+ARG VECTORSCAN_SHA=71fae7ee8d63e1513a6df762cdb5d5f02a9120a2422cf1f31d57747c2b8d36ab
+
 ################################################################################
-# Build Nosey Parker
+# Base stage
 ################################################################################
-FROM rust:1.65 as build
+FROM rust:$RUST_VER AS base
+
+ARG VECTORSCAN_VER
+ARG VECTORSCAN_SHA
+
+ENV HYPERSCAN_ROOT "/vectorscan/build"
+
+WORKDIR "/vectorscan"
+
+ADD https://github.com/VectorCamp/vectorscan/archive/refs/tags/vectorscan/$VECTORSCAN_VER.tar.gz ./vectorscan.tar.gz
 
 # Install dependencies
-RUN apt-get update && apt-get install -y \
+RUN apt-get update &&\
+    apt-get install -y \
     build-essential \
     cmake \
     git \
     libboost-dev \
     ninja-build \
     pkg-config \
-    ragel
-
-# Build vectorscan from source
-WORKDIR "/vectorscan"
-ADD https://github.com/VectorCamp/vectorscan/archive/refs/tags/vectorscan/5.4.8.tar.gz vectorscan.tar.gz
-RUN echo '71fae7ee8d63e1513a6df762cdb5d5f02a9120a2422cf1f31d57747c2b8d36ab vectorscan.tar.gz' | sha256sum -c && \
-    tar --strip-components 1 -xzf vectorscan.tar.gz && \
-    cmake -S . -B build -GNinja -DCMAKE_BUILD_TYPE=Release -DFAT_RUNTIME=OFF && \
+    ragel &&\
+    apt-get clean &&\
+    # Build vectorscan from source
+    echo "$VECTORSCAN_SHA vectorscan.tar.gz" | sha256sum -c &&\
+    tar --strip-components 1 -xzf vectorscan.tar.gz &&\
+    rm -rf vectorscan.tar.gz &&\
+    cmake -S . -B build -GNinja -DCMAKE_BUILD_TYPE=Release -DFAT_RUNTIME=OFF &&\
     cmake --build build
 
-# Build Nosey Parker
+################################################################################
+# Build Rust dependencies, caching stage
+################################################################################
+FROM base AS dependencies
+
 WORKDIR "/noseyparker"
+
+COPY ["Cargo.toml", "Cargo.lock",  "./"]
+
+# Make directory structure
+RUN mkdir -p  ./src/bin/noseyparker &&\
+    mkdir -p ./benches &&\
+    # Create stub files for build
+    touch ./benches/microbench.rs ./src/lib.rs &&\
+    echo "fn main() {}" > ./src/bin/noseyparker/main.rs &&\
+    # Run the build
+    cargo build --bin noseyparker --release
```

Comment:
I _think_ I understand what's going on here: all this song and dance is to convince `cargo` to build just the dependencies of the project, essentially, which it doesn't support natively. If I understand that right, please add a 1-2 sentence comment noting what's going on, for future readers.

Is it actually necessary to create dummy `benches/microbench.rs` and `src/lib.rs`, or is it sufficient to only create the dummy `src/bin/noseyparker/main.rs`?

Reply:
Added comments in an attempt to describe what is going on. Those three files are needed to do a build even if you specify just the bin target unfortunately.

Reply:
Thanks!